Feature: Model Export and Import

mikeroySoft opened the pull request:
**Added Model Export and Import  #features.**

Finally had time to revisit this issue: https://github.com/ollama/ollama/issues/335 

This feature allows users to easily backup and transfer models between Ollama installations via USB drives or network storage.

- Added `export` command to copy models to file/directory
- Added `import` command to load models from file/directory
- Supports directory, tar, and tar.gz formats
- Includes checksum verification and progress tracking
- Added --compress flag for export (this is slow, but shaves off around 2% and creates a .tgz or .tar.gz file)
- Added --force flag for import to override an existing model
- Added /--insecure flags to skip verification during import

**Usage**

**Export **

`ollama export  MODEL DESTINATION [flags]`

Flags:
      `--compress`       Compress the export as tar.gz or .tgz (slower, slightly smaller file size, single-file export)
      `-h`, `--help`       Help for export

**Import**
`ollama import SOURCE [MODEL_NAME] [flags]`

Flags:
      `--force`               Overwrite existing model
      `-h`, `--help`       Help for import
      `--insecure`         Skip checksum verification

jcheek commented:
exporting from Mac M3 appears to work.  Will test importing next.

```
Macximum (ziquid/add-ie-docs):~/scm/ollama$ go run . export magi-40k:latest ~/magi --compress
# github.com/ollama/ollama
ld: warning: ignoring duplicate libraries: '-lobjc'
exporting magi-40k:latest
exporting ade4614f24e1
exporting 7a14fc708c00
exporting d7be94e9f19c
exporting 43070e2d4e53
exporting 71442892dfbd
exporting ba3ab63d3582
export complete
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ dir ~/magi
Permissions Size User   Date Modified Name
.rw-r--r--@  14G joseph 22 Jun 15:46   /Users/joseph/magi
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ file ~/magi
/Users/joseph/magi: gzip compressed data, original size modulo 2^32 1449032704
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ tar tzf ~/magi
metadata.json
manifest.json
blobs/sha256-ade4614f24e1529c492da184b0572770497303ef2bcfcd3a5ec334fca8354ec3
blobs/sha256-7a14fc708c00c0db687494a1e2748e61af930ba3248e4b91d383ff2a4b5e899f
blobs/sha256-d7be94e9f19c39440923a6f3a7f489fb2779d00941cf2b5bcc8abf1f4a7aba00
blobs/sha256-43070e2d4e532684de521b885f385d0841030efa2b1a20bafb76133a5e1379c1
blobs/sha256-71442892dfbd23ccfa13184132f930d8ac1d07763236baae97fc606095984ea3
blobs/sha256-ba3ab63d35822e9df884ce427fa8f553655296324e035ccf79523e1e293fbd9b
```

jcheek commented:
import is not working for me.  On Mac M3 (same system as export happened):

```
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi magi-import-test
# github.com/ollama/ollama
ld: warning: ignoring duplicate libraries: '-lobjc'

Error: unsupported import format
exit status 1
```

Note that I exported the model using `--compress` but didn't give it a .tar.gz or .tgz extension.  `--import` doesn't seem to realize that it's a tarball.

On x86_64 Linux (AlmaLinux 9.6), CUDA 12.9 (~/magi was scp'ed to this host):

```
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi

Error: unmarshal: invalid character 'p' after top-level value
exit status 1
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi --insecure

Error: unmarshal: invalid character 'p' after top-level value
exit status 1
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi magi40k-latest

Error: unmarshal: invalid character 'p' after top-level value
exit status 1
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi magi40k-latest --insecure

Error: unmarshal: invalid character 'p' after top-level value
exit status 1
```

jcheek commented:
On Mac M3, renaming the file to have a `.tgz` extension allowed it to import.

```
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ mv ~/magi ~/magi.tgz
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi.tgz magi-import-test
# github.com/ollama/ollama
ld: warning: ignoring duplicate libraries: '-lobjc'
validating import package
importing magi-import-test
skipping existing ade4614f24e1
skipping existing 7a14fc708c00
skipping existing d7be94e9f19c
skipping existing 43070e2d4e53
skipping existing 71442892dfbd
skipping existing ba3ab63d3582
import complete
```

Both ollama 0.92 and this fork showed the import:

```
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ ollama ls | grep magi
magi-import-test:latest                                         73bafbf7de66    14 GB     3 minutes ago
magi-40k:latest                                                 73bafbf7de66    14 GB     3 hours ago
huihui_ai/magistral-abliterated:latest                          cc063d4f829d    14 GB     4 hours ago
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ go run . ls | grep magi
# github.com/ollama/ollama
ld: warning: ignoring duplicate libraries: '-lobjc'
magi-import-test:latest                                         73bafbf7de66    14 GB     3 minutes ago
magi-40k:latest                                                 73bafbf7de66    14 GB     3 hours ago
huihui_ai/magistral-abliterated:latest                          cc063d4f829d    14 GB     4 hours ago
```

Both ollama 0.92 and this fork ran the newly-imported model:

```
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ go run . run magi-import-test
# github.com/ollama/ollama
ld: warning: ignoring duplicate libraries: '-lobjc'
>>> hello
Greetings! I see you've just entered a quaint, candlelit study, [snip]

>>> Send a message (/? for help)
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ ollama run magi-import-test 'hello'
The sun hung low in the sky, casting a warm golden glow over the cobblestone streets of the old town.  [snip]
You took a deep breath, savoring the moment before turning your attention back to the user who had just greeted
you—"hello."
```

(Note: I did stop and start the ollama server each time to ensure I was using the correct server with the matching client).

jcheek commented:
> Does it succeed if you either rename the file with a tar.gz extension OR if you specify the tar.gz or .tgz during the export command?

On Mac, yes.  Linux, no.

jcheek commented:
On Linux x86_64/CUDA 12.9, the model did not import after renaming it to end with a `.tgz` extension.  Still getting the same error, `Error: unmarshal: invalid character 'p' after top-level value`.

Possible architecture error, ARM vs x86?  I'm not sure.

```
asrock (ziquid/add-ie-docs=):~/scm/ollama$ mv ~/magi ~/magi.tgz
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi.tgz

Error: unmarshal: invalid character 'p' after top-level value
exit status 1
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi.tgz
asrock (ziquid/add-ie-docs=):~/scm/ollama$ file ~/magi.tgz
/home/joseph/magi.tgz: gzip compressed data, original size modulo 2^32 1449032704
asrock (ziquid/add-ie-docs=):~/scm/ollama$ tar tzf ~/magi.tgz
metadata.json
manifest.json
blobs/sha256-ade4614f24e1529c492da184b0572770497303ef2bcfcd3a5ec334fca8354ec3
blobs/sha256-7a14fc708c00c0db687494a1e2748e61af930ba3248e4b91d383ff2a4b5e899f
blobs/sha256-d7be94e9f19c39440923a6f3a7f489fb2779d00941cf2b5bcc8abf1f4a7aba00
blobs/sha256-43070e2d4e532684de521b885f385d0841030efa2b1a20bafb76133a5e1379c1
blobs/sha256-71442892dfbd23ccfa13184132f930d8ac1d07763236baae97fc606095984ea3
blobs/sha256-ba3ab63d35822e9df884ce427fa8f553655296324e035ccf79523e1e293fbd9b
```

jcheek commented:
MD5sums of the files match:

```
Macximum (ziquid/add-ie-docs=):~/scm/ollama$ md5 ~/magi.tgz
MD5 (/Users/joseph/magi.tgz) = d1029ccd86e34e918fe198f622d590cc
asrock (ziquid/add-ie-docs=):~/scm/ollama$ md5sum ~/magi.tgz
d1029ccd86e34e918fe198f622d590cc  /home/joseph/magi.tgz
```

I'll try a linux export/import to see if that works.

jcheek commented:
https://github.com/ollama/ollama/issues/2293 leads me to believe that I have a mismatched ollama client and server on my linux box.  Trying again.

jcheek commented:
Yes, that was it!  In my haste I had forgotten to kill the existing ollama 0.92 server running on my Linux box.  Now:

```
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run  . serve
[lots of debug snipped]
```

and

```
asrock (ziquid/add-ie-docs=):~/scm/ollama$ go run . import ~/magi.tgz
validating import package
importing magi-40k:latest
importing ade4614f24e1
importing 7a14fc708c00
importing d7be94e9f19c
importing 43070e2d4e53
importing 71442892dfbd
importing ba3ab63d3582
import complete
```

mikeroySoft commented:
> How about adding zstd as default compress algo, it has better compression rate and speed.
Nice... gave it a whirl and had very happy results. It's default now =)
I also added parallelism to the export, so it's quicker, and also streams to file vs. using in-memory buffers (because models are big...)
Also added the zstd parameter for compression strength. Using level 19 I took a 19.85GB model (deepseek-r2:32b) down to 19.66GB. 
Doesn't seem to add much, but I'm sure every little bit helps when dealing with 100GB+ models, and it doesn't seem to take too long vs. uncompressed.

jcheek commented on README.md:
````diff
@@ -175,6 +175,50 @@ ollama rm llama3.2
+
+```shell
+# Import a model directory
+ollama import /~/my-models/llama4-model-backup
````
errant initial / ?  Maybe you meant

ollama import ~/my-models/llama4-model-backup

???

mikeroySoft replied:
Nice catch! My typo entirely... new commit just pushed

jcheek commented:
Kudos for adding docs, btw.  @mikeroySoft++

mikeroySoft commented:
Did a bunch of cleanup and added some sanity checks to make sure we don't overwrite files on export. Added a `--force` flag to override that as well. I think it's in pretty good shape for a review =)

mikeroySoft commented:
Went through the feature and added integration and functional testing as per Contribution guidelines. 

If anyone wanted the extensive 'how this feature and its tests was built' detail, the files are attached:
[export-import-testing-summary.md](https://github.com/user-attachments/files/21201894/export-import-testing-summary.md)
[export-import-design.md](https://github.com/user-attachments/files/21201895/export-import-design.md)
[export-import-plan.md](https://github.com/user-attachments/files/21201896/export-import-plan.md)
